pkg/kubelet/kubelet_test.go

```diff
@@ -573,6 +574,111 @@ func TestHandleMemExceeded(t *testing.T) {
 	checkPodStatus(t, kl, fittingPod, v1.PodPending)
 }
 
+// pluginResourceSpec is stub of a resource registered in device plugin handler
+// including its quantity and expected result of allocation
+type pluginResourceSpec struct {
+	quantity        resource.Quantity
+	allocateSucceed bool
+}
+
+// Tests that we handle result of interface UpdatePluginResources correctly
+// by setting corresponding status in status map.
+func TestHandlePluginResources(t *testing.T) {
+	testKubelet := newTestKubelet(t, false /* controllerAttachDetachEnabled */)
+	defer testKubelet.Cleanup()
+	testKubelet.chainMock()
+	kl := testKubelet.kubelet
+
+	adjustedResource := v1.ResourceName("domain1.com/adjustedResource")
+	unadjustedResouce := v1.ResourceName("domain2.com/unadjustedResouce")
+	failedResource := v1.ResourceName("domain2.com/failedResource")
+	resourceQuantity1 := *resource.NewQuantity(int64(1), resource.DecimalSI)
+	resourceQuantity2 := *resource.NewQuantity(int64(2), resource.DecimalSI)
+	sufficientQuantity := *resource.NewQuantity(int64(40), resource.DecimalSI)
+	resourcesStub := map[v1.ResourceName]pluginResourceSpec{
+		adjustedResource:  {quantity: resourceQuantity1, allocateSucceed: true},
+		unadjustedResouce: {quantity: resourceQuantity1, allocateSucceed: true},
+		failedResource:    {quantity: resourceQuantity1, allocateSucceed: false},
+	}
+	nodes := []*v1.Node{
+		{ObjectMeta: metav1.ObjectMeta{Name: testKubeletHostname},
+			Status: v1.NodeStatus{Capacity: v1.ResourceList{}, Allocatable: v1.ResourceList{
+				// assign sufficient resources in nodeInfo, to ensure test results
+				// not affected by other parts of Admission
+				adjustedResource:  sufficientQuantity,
+				unadjustedResouce: sufficientQuantity,
+				failedResource:    sufficientQuantity,
+				v1.ResourcePods:   sufficientQuantity,
+			}}},
+	}
+	kl.nodeInfo = testNodeInfo{nodes: nodes}
+
+	updatePluginResourcesFunc := func(node *schedulercache.NodeInfo, attrs *lifecycle.PodAdmitAttributes) error {
```

Comment:
I am a bit confused by the logic here. How do we handle adjustedResource and unadjustedResouce differently in updatePluginResourcesFunc()? Did you intend to set node.allocatable[] = resourceSpec.quantity here? But I thought the intention is to adjust node allocatable for adjustedResource and not adjust node allocatable for unadjustedResouce. Suggest we simplify the test logic as below:

Reply:
I think I saw the problem you ran into. Basically, the default PredicateAdmitHandler we added in newTestKubelet() is NOT a no-op but still runs the GeneralPredicate logic. But instead of set all resource with sufficientQuantity, I would suggest we just reset kl.admitHandlers before adding a new PredicateAdmitHandler. Another problem I noticed in this test is that we only set resource Limits in all PodSpecs, but in scheduler predicate code, I think it actually looks at resource Requests instead, so we are not really testing the behavior we want to test here. For extended resources, we expect Limits and Requests to always be equal, so I think we should change the code here to set both. I changed this test to include these changes. Could you take a look and merge it with this PR? 
https://github.com/jiayingz/kubernetes/blob/pr54488/pkg/kubelet/kubelet_test.go#L579

Reply:
> Another problem I noticed in this test is that we only set resource Limits in all PodSpecs, but in scheduler predicate code, I think it actually looks at resource Requests instead, so we are not really testing the behavior we want to test here

that makes sense

>I would suggest we just reset kl.admitHandlers before adding a new PredicateAdmitHandler

The method itself is okey. However, I thought one purpose of the test here is to ensure the new updatePluginResources func works fine with other parts in PredicateAdmitHandler. I feel it kind of like a test of DevicePluginHander if we set admitHandlers empty here.